src/lib/db/api-token-store.ts

```diff
@@ -72,26 +97,64 @@ export class ApiTokenStore implements IApiTokenStore {
 
     async getAll(): Promise<IApiToken[]> {
         const stopTimer = this.timer('getAll');
-        const rows = await this.db<ITokenTable>(TABLE);
+        const rows = await this.makeTokenProjectQuery();
         stopTimer();
-        return rows.map(toToken);
+        return toTokens(rows);
     }
 
     async getAllActive(): Promise<IApiToken[]> {
         const stopTimer = this.timer('getAllActive');
-        const rows = await this.db<ITokenTable>(TABLE)
+        const rows = await this.makeTokenProjectQuery()
             .where('expires_at', 'IS', null)
             .orWhere('expires_at', '>', 'now()');
         stopTimer();
-        return rows.map(toToken);
+        return toTokens(rows);
+    }
+
+    private makeTokenProjectQuery() {
+        return this.db<ITokenRow>(`${TABLE} as tokens`)
+            .leftJoin(
+                `${API_LINK_TABLE} as token_project_link`,
+                'tokens.secret',
+                'token_project_link.secret',
+            )
+            .select(
+                'tokens.secret',
+                'username',
+                'type',
+                'expires_at',
+                'created_at',
+                'seen_at',
+                'environment',
+                'token_project_link.project',
+            );
     }
 
     async insert(newToken: IApiTokenCreate): Promise<IApiToken> {
-        const [row] = await this.db<ITokenTable>(TABLE).insert(
-            toRow(newToken),
-            ['created_at'],
-        );
-        return { ...newToken, createdAt: row.created_at };
+        const response = await this.db.transaction(async (tx) => {
+            const [row] = await tx<ITokenInsert>(TABLE).insert(
+                toRow(newToken),
+                ['created_at'],
+            );
+
+            const updateProjectTasks = (newToken.projects || [])
+                .filter((project) => {
+                    return project !== ALL_PROJECTS;
+                })
+                .map((project) => {
+                    return tx.raw(
+                        `INSERT INTO ${API_LINK_TABLE} VALUES (?, ?)`,
+                        [newToken.secret, project],
+                    );
+                });
+            await Promise.all(updateProjectTasks);
```

Comment:
Yo do not need to create a bunch of insert request. You just need to to map the projects in it to a list of "objects" matching the "API_LINK_TABLE" shape and perform a simple knex insert on all the values.

Reply:
What Ivar said, knex already has efficient batch insert, write the query, give a list of tuples with the same amount of args as your sql insert. See src/lib/db/clientMetricsStore-v2.ts:101 for instance.

Though this seems to have the advantage that it's transactional and as such will rollback the LINK_TABLE inserts if any of them fail.

Reply:
that can be achieved anyway, and yes the transaction stuff is the important part here!